Element: Fix serializer handling of multiple distinct contexts

aduth opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/pull/20721#discussion_r397072949
Previously: #8189, #5897

This pull request seeks to resolve an issue with the implementation of the custom `@wordpress/element` serializer where the a context consumer is only able to consume the value of the closest context provider, even if it is not of the same context type. As can be observed in the included test case, the result was that two distinct context providers would not interoperate.

**Implementation Notes:**

As described at https://github.com/WordPress/gutenberg/pull/20721#discussion_r398075725, the underlying problem is that when a provider renders its children, it _replaces_ the current context recursion argument, which prevents multiple provider values from coexisting in the same element tree.

The proposed changes seek to resolve this by changing the context argument to a `Map` object, where a provider extends the map with its value. At the time the consumer is rendered, it performs a lookup of the provider's value using the components internal `type._context.Provider` value.

**Testing Instructions:**

Ensure unit tests pass:

```
npm run test-unit
```

aduth committed:
Element: Fix serializer handling of multiple distinct contexts

gziolo commented on packages/element/src/test/serialize.js:
```diff
@@ -317,6 +317,36 @@ describe( 'renderElement()', () => {
+			Provider: SecondProvider,
+		} = createContext();
+
+		const result = renderElement(
```
I see that e2e tests fail for blocks that have nesting. I'm wondering if there should be another unit tests that replicate how `InnerBlocks` works, if I follow it properly it would be something like:
```js
<FirstProvider value="First">
 	<FirstConsumer>
 		{ ( first ) => (
 			<SecondProvider value="Second">
 				<SecondConsumer>
 					{ ( second ) => `First: { first }, Second: ${ second }` }
 				</SecondConsumer>
			</SecondProvider>
 		) }
	</FirstConsumer>
</FirstProvider>
```

Although, I don't see any reason why it would make any difference.

aduth replied:
I'm not really sure what's going on with the end-to-end tests, to be honest. I've not been able to reproduce those same failures in my local environment.

aduth replied:
I added a test case for this in 90dacf058 for good measure. It passes, as expected. Still struggling with the end-to-end tests in Travis.

gziolo replied:
I've just realized that we use one shared context for `InnerBlocks` so the following might be closer to the use case that fails on Travis:

```js
<Provider value="outer">
 	<Consumer>
 		{ ( outter ) => (
 			< Provider value="inner">
 				<Consumer>
 					{ ( inner ) => `Outer: { outer }, Inner: ${ inner }` }
 				</Consumer>
			</Provider>
 		) }
	</Consumer>
</Provider>
```

I don't think it's much different from a similar existing test though. I can't think of any reason why it fails on Travis :(

ellatrix replied:
Oh no :(

aduth committed:
Element: Add non-mixed test case for nested context

aduth committed:
Element: Add testt case for rendering Context directly

Not officially valid, yet technically supported in React's implementation.

See: https://github.com/facebook/react/blob/a16b34974508cd23ce0844ad09a0e37a879d5591/packages/react-dom/src/server/ReactPartialRenderer.js#L1158-L1160

aduth committed:
Element: Guard property access to context consumer `_context`

Multiple references in React's codebase check whether this is assigned.

See: https://github.com/facebook/react/blob/a16b34974508cd23ce0844ad09a0e37a879d5591/packages/react-dom/src/server/ReactPartialRenderer.js#L1157

It's not entirely clear the exact circumstances this may occur at this point in code. Ideally there would be a test case verifying failure without this guard.

aduth committed:
Try common Map implementation

aduth committed:
Try avoiding explicit argument to constructor

ellatrix commented on packages/element/src/serialize.js:
```diff
@@ -36,6 +36,7 @@ import {
 	kebabCase,
 	isPlainObject,
 } from 'lodash';
+import CoreJSMap from 'core-js-pure/features/map';
```
Any reason why we use this?

aduth replied:
It's meant to be temporary, an experiment to see if it would fix the previously-observed issue to use a very predictable implementation.
